comp/core/tagger/taggerimpl/server/server.go

```diff
@@ -27,15 +28,30 @@ const (
 	streamKeepAliveInterval = 9 * time.Minute
 )
 
+// streamIDManager is used to generate unique ID's for incoming streaming requests
+// This unique ID is used to subscribe to the tagger
+// TODO: remove this struct when the protobuf of the stream request is updated to use filters.
+type streamIDManager int32
+
+func (s *streamIDManager) getNextUniqueID() string {
+	id := fmt.Sprintf("stream-client-%d", *s)
+	atomic.AddInt32((*int32)(s), 1)
```

Comment:
If the `streamIDManager` field is meant to be modified and accessed concurrently, it should use a safe `sync/atomic` type.

Using a “generic standard” type like `int32` leads to concurrency bugs like the one in the current implementation of `getNextUniqueID`:
* Whereas the increment done line 38 is thread-safe thanks to the use [`atomic.AddInt32`](https://pkg.go.dev/sync/atomic#AddInt32),
* The read done line 37 is **not** thread-safe because of the lack of use of [`atomic.LoadInt32`]( https://pkg.go.dev/sync/atomic#LoadInt32).

This first bug could be fixed by:
```suggestion
type streamIDManager int32

func (s *streamIDManager) getNextUniqueID() string {
	id := fmt.Sprintf("stream-client-%d", atomic.LoadInt32(s))
	atomic.AddInt32((*int32)(s), 1)
```

However, in order to avoid this kind of bug to be possible, it would be better to use a dedicated type that prevents non-atomic accesses:
```suggestion
type streamIDManager struct {
	atomic.Int64
}

func (s *streamIDManager) getNextUniqueID() string {
	id := fmt.Sprintf("stream-client-%d", s.Load())
	s.Add(1)
```

Although all accesses (both read and writes) to the `streamIDManager` field are now thread-safe, the `getNextUniqueID` function still suffer from a race condition.
It’s still possible to have two threads T1 and T2 calling `getNextUniqueID` and return exactly the same value with the following sequence:
```
  | value
  | of  s                      thread T1                                         thread T2
  |
  |  42
  |           id := fmt.Sprintf("stream-client-%d", s.Load())
  |  42
  |                                                             id := fmt.Sprintf("stream-client-%d", s.Load())
  |  42
  |           s.Add(1)
  |  43
  |                                                             s.Add(1)
  |  44
  |           return id
  |  44
  |                                                             return id
  |  44
  |
  V

time
```

With such an execution flow, with `s` being `42` before `getNextUniqueID` is being concurrently called by the two threads T1 and T2, `getNextUniqueID` would return exactly the same ID (`stream-client-42`) in both threads. Meaning that it wouldn’t be unique.
The following call would return `stream-client-44`.

It’s crucial that the read of the `s` value and its increment are atomic, which is not the case with any of the proposition above.

The most generic way to make the `fmt.Sprintf(…, s.Load())` and the `s.Add(1)` lines atomic is to protect them by a mutex.
But in this precise case, it would be overkill as we can atomically add and load the `s` value with:
```suggestion
type streamIDManager struct {
	atomic.Int64
}

func (s *streamIDManager) getNextUniqueID() string {
	id := fmt.Sprintf("stream-client-%d", s.Add(1))
```

Reply:
Awesome explanation !!! 🔥🔥🔥🔥🔥🔥

Thanks @L3n41c !!